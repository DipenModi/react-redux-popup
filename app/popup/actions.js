export const OPEN_POPUP = 'OPEN_POPUP';
export const CLOSE_POPUP = 'CLOSE_POPUP';

export function openPopup(popupId, container) {
    return {
        container,
        popupId,
        type: OPEN_POPUP
    };
}

export function closePopup(popupId) {
    return {
        popupId,
        type: CLOSE_POPUP
    }
}

export default { openPopup, closePopup };
