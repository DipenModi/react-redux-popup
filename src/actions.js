export const OPEN_POPUP = 'OPEN_POPUP';
export const CLOSE_POPUP = 'CLOSE_POPUP';
export const UPDATE_POPUP_PROPS = 'UPDATE_POPUP_PROPS';
export const UPDATE_POPUP_SCROLL_POSITION = 'UPDATE_POPUP_SCROLL_POSITION';

export function openPopup(popupId, rect) {
    return {
        rect,
        popupId,
        type: OPEN_POPUP
    };
}

export function closePopup(popupId) {
    return {
        popupId,
        type: CLOSE_POPUP
    }
}

export function updatePopupProps(popupId, props) {
    return {
        popupId,
        props,
        type: UPDATE_POPUP_PROPS
    };
}

export function updateScrollPosition(x, y) {
    return {
        x,
        y,
        type: UPDATE_POPUP_SCROLL_POSITION
    };
}
