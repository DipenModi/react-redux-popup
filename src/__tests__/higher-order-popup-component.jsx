import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import ConnectedComponent, { HigherOrderPopupComponent } from 'rrp/higher-order-popup-component';
import collection, { clearAll } from 'rrp/popup-collection';
import { updatePopupProps, closePopup } from 'rrp/actions';

describe('higher-order-popup-component', function() {
    const origConsoleError = console.error;
    const handler = () => true;
    const mockStore = configureStore();
    const ComposedComponent = <div>Test</div>;
    const TestComponent = HigherOrderPopupComponent(ComposedComponent, 'someType');

    beforeEach(clearAll);
    afterEach(function() {
        // eslint-disable-next-line no-console
        console.error = origConsoleError;

        clearAll();
    });

    it('should render the component', function() {
        const wrapper = shallow(
            <TestComponent
                closePopup={handler}
                id="testPopup"
                prop1="prop1_value"
                updatePopupProps={handler}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should render the component with provider', function() {
        const ConnectedTestComponent = ConnectedComponent(ComposedComponent, 'someType');
        const wrapper = shallow(
            <Provider store={mockStore()}>
                <ConnectedTestComponent
                    closePopup={handler}
                    id="testPopup"
                    prop1="prop1_value"
                    updatePopupProps={handler}
                />
            </Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('should push and remove from collection when mounted and unmounted', function() {
        const closePopup = jest.fn();
        const wrapper = mount(
            <TestComponent
                closePopup={closePopup}
                id="testPopup"
                prop1="prop1_value"
                updatePopupProps={handler}
            />
        );
        expect(collection.length).toBe(1);
        expect(collection).toEqual(expect.arrayContaining([
            [
                'someType',
                ComposedComponent,
                expect.objectContaining({
                    prop1: 'prop1_value'
                })
            ]
        ]));

        wrapper.unmount();
        expect(collection.length).toBe(0);
        expect(closePopup).toHaveBeenCalledWith('testPopup');
    });

    it('should dispatch closePopup action when closePopup is not specified', function() {
        const dispatch = jest.fn();
        const wrapper = mount(
            <TestComponent
                dispatch={dispatch}
                id="testPopup"
                prop1="prop1_value"
                updatePopupProps={handler}
            />
        );
        wrapper.unmount();
        expect(dispatch).toHaveBeenCalledTimes(1);
        expect(dispatch).toHaveBeenCalledWith(closePopup('testPopup'));
    });

    it('should update the props in the collection and call updatePopupProps', function() {
        const updatePopupProps = jest.fn();
        const wrapper = mount(
            <TestComponent
                closePopup={handler}
                id="testPopup"
                prop1="prop1_value"
                updatePopupProps={updatePopupProps}
            />
        );
        wrapper.setProps({ prop1: 'updated_value' });
        expect(updatePopupProps).toHaveBeenCalledWith('testPopup', {
            updatePopupProps,
            closePopup: handler,
            id: 'testPopup',
            prop1: 'updated_value'
        });

        wrapper.unmount();
    });

    it('should not update the props in the collection and call updatePopupProps when props are same', function() {
        const updatePopupProps = jest.fn();
        const wrapper = mount(
            <TestComponent
                closePopup={() => true}
                id="testPopup"
                prop1="prop1_value"
                updatePopupProps={updatePopupProps}
            />
        );
        wrapper.update();
        expect(updatePopupProps).toHaveBeenCalledTimes(0);
        wrapper.unmount();
    });

    it('should warn the missing id property', function() {
        // eslint-disable-next-line no-console
        console.error = jest.fn();
        shallow(
            <TestComponent
                closePopup={handler}
                prop1="prop1_value"
                updatePopupProps={handler}
            />
        );

        // eslint-disable-next-line no-console
        expect(console.error).toHaveBeenCalled();
    });
});
