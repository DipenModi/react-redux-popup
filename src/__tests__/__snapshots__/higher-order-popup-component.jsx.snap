// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`higher-order-popup-component should render the component 1`] = `null`;

exports[`higher-order-popup-component should render the component with provider 1`] = `
<Connect(PopupComponent)
  closePopup={[Function]}
  id="testPopup"
  prop1="prop1_value"
  updatePopupProps={[Function]}
/>
`;
